test/NuGet.Core.Tests/NuGet.Configuration.Test/SearchTreeTests.cs

```diff
@@ -0,0 +1,344 @@
+
+using System.IO;
+using System.Linq;
+using FluentAssertions;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace NuGet.Configuration.Test
+{
+    public class SearchTreeTests
+    {
+        [Fact]
+        public void SearchTree_WithOneSource()
+        {
+            // Arrange
+            using var mockBaseDirectory = TestDirectory.Create();
+            var configPath1 = Path.Combine(mockBaseDirectory, "NuGet.Config");
+            SettingsTestUtils.CreateConfigurationFile(configPath1, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <clear />
+        <packageSource key=""nuget.org"">
+            <namespace id=""stuff"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var settings = Settings.LoadSettingsGivenConfigPaths(new string[] { configPath1 });
+
+            // Act & Assert
+            var configuration = PackageNamespacesConfiguration.GetPackageNamespacesConfiguration(settings);
+            Assert.True(configuration.IsNamespacesEnabled);
+
+            configuration.Namespaces.Should().HaveCount(1);
+            var packageSourcesMatchFull = configuration.GetConfiguredPackageSources("stuff");
+            Assert.Equal(1, packageSourcesMatchFull.Count);
+            Assert.Equal("nuget.org", packageSourcesMatchFull.First());
+
+            var packageSourcesMatchPartial = configuration.GetConfiguredPackageSources("stu");
+            Assert.Null(packageSourcesMatchPartial);
+
+            var packageSourcesNoMatch = configuration.GetConfiguredPackageSources("random");
+            Assert.Null(packageSourcesNoMatch);
+        }
+
+        [Fact]
+        public void SearchTree_WithOneSourceMultipart()
+        {
+            // Arrange
+            using var mockBaseDirectory = TestDirectory.Create();
+            var configPath1 = Path.Combine(mockBaseDirectory, "NuGet.Config");
+            SettingsTestUtils.CreateConfigurationFile(configPath1, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <clear />
+        <packageSource key=""PublicRepository"">
+            <namespace id=""Contoso.Opensource.*"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var settings = Settings.LoadSettingsGivenConfigPaths(new string[] { configPath1 });
+
+            // Act & Assert
+            var configuration = PackageNamespacesConfiguration.GetPackageNamespacesConfiguration(settings);
+            Assert.True(configuration.IsNamespacesEnabled);
+            configuration.Namespaces.Should().HaveCount(1);
+
+            // No match
+            var packageSourcesMatchPartial1 = configuration.GetConfiguredPackageSources("Cont");
+            Assert.Null(packageSourcesMatchPartial1);
+
+            // No match
+            var packageSourcesMatchPartial2 = configuration.GetConfiguredPackageSources("Contoso.Opensource");
```

Comment:
Please add a test for `Contoso.Opensource.`

Reply:
https://github.com/NuGet/NuGet.Client/blob/42e68a74fce207a2b9c442886071f51d2c20ebe5/test/NuGet.Core.Tests/NuGet.Configuration.Test/SearchTreeTests.cs#L81-L83

Reply:
I added following test. Considering `*` as zero or more currently it matches with `Contoso.Opensource.*`. Should it be null?
https://github.com/NuGet/NuGet.Client/blob/91f571d40101d0b0b836b7dece071799f85aaa20/test/NuGet.Core.Tests/NuGet.Configuration.Test/SearchTreeTests.cs#L81-L84

Reply:
@nkolev92 
It's kind of edge case. Maybe we just reject this input right away due to being invalid package Id. What is your opinion?

Reply:
Actually we only check if `package Id` is valid when we create time but we don't at restore. So maybe `SearchTree` is not good place to check validity of `package Id`. Even SearchTree doesn't fail on it, eventually it'll fail as it was before this PR.
So testing `Contoso.Opensource.` is not really relevant here.

Reply:
I just manually created `Foo..1.0.0.nupkg` and successfully restored with `nuget.exe 5.9 GA`. So above assertion is valid even though package Id seems invalid.

Reply:
Yeah, I see there was a teams discussion as well. 

`NuGet*` matches `NuGet`. 

Resolving this thread.